core/src/banking_stage/latest_unprocessed_votes.rs

```diff
@@ -283,6 +300,48 @@ impl LatestUnprocessedVotes {
             .and_then(|l| l.read().unwrap().timestamp())
     }
 
+    #[cfg(test)]
+    pub(crate) fn set_staked_nodes(&self, staked_nodes: &[Pubkey]) {
+        let staked_nodes: HashMap<Pubkey, u64> =
+            staked_nodes.iter().map(|pk| (*pk, 1u64)).collect();
+        *self.cached_staked_nodes.write().unwrap() = Arc::new(staked_nodes);
+    }
+
+    fn weighted_random_order_by_stake(&self) -> impl Iterator<Item = Pubkey> {
+        // Efraimidis and Spirakis algo for weighted random sample without replacement
+        let staked_nodes = self.cached_staked_nodes.read().unwrap();
+        let latest_votes_per_pubkey = self.latest_votes_per_pubkey.read().unwrap();
+        let mut pubkey_with_weight: Vec<(f64, Pubkey)> = latest_votes_per_pubkey
+            .keys()
+            .filter_map(|&pubkey| {
+                let stake = staked_nodes.get(&pubkey).copied().unwrap_or(0);
+                if stake == 0 {
+                    None // Ignore votes from unstaked validators
+                } else {
+                    Some((thread_rng().gen::<f64>().powf(1.0 / (stake as f64)), pubkey))
+                }
+            })
+            .collect::<Vec<_>>();
+        pubkey_with_weight.sort_by(|(w1, _), (w2, _)| w1.partial_cmp(w2).unwrap());
+        pubkey_with_weight.into_iter().map(|(_, pubkey)| pubkey)
+    }
+
+    /// Recache the staked nodes based on a bank from the new epoch.
+    /// This should only be run by the TPU vote thread
```

Comment:
Why should this only be run by the TPU vote thread?

Reply:
there's no need for both threads to compete for this write lock. Since the TPU vote thread is the only consumer of this data structure we set it up so it also updates the epoch info